pkg/clientutil/volumes.go

```diff
@@ -0,0 +1,167 @@
+			}
+		}
+	}
+
+	return nil
+}
+
+func disablePVCreation(ctx context.Context, c client.Client) error {
+	// Prevent re-creation of PVs and PVCs by using an intentionally defunct admissionWebhook
+	creatorGetters := []reconciling.NamedValidatingWebhookConfigurationReconcilerFactory{
+		creationPreventingWebhook("", VolumeResources),
+	}
+	if err := reconciling.ReconcileValidatingWebhookConfigurations(ctx, creatorGetters, "", c); err != nil {
+		return fail.KubeClient(err, "failed to create ValidatingWebhookConfiguration to prevent creation of PVs/PVCs.")
+	}
+
+	return nil
+}
+
+func cleanupPVCUsingPods(ctx context.Context, c client.Client) error {
+	podList := &corev1.PodList{}
+	if err := c.List(ctx, podList); err != nil {
+		return fail.KubeClient(err, "failed to list Pods from user cluster.")
+	}
+
+	var pvUsingPods []*corev1.Pod
+	for idx := range podList.Items {
+		pod := &podList.Items[idx]
+		if podUsesPV(pod) {
+			pvUsingPods = append(pvUsingPods, pod)
+		}
+	}
+
+	for _, pod := range pvUsingPods {
+		if pod.DeletionTimestamp == nil {
+			if err := DeleteIfExists(ctx, c, pod); err != nil {
+				return fail.KubeClient(err, "failed to delete Pod.")
+			}
+		}
+	}
+
+	return nil
+}
+
+func podUsesPV(p *corev1.Pod) bool {
+	for _, volume := range p.Spec.Volumes {
+		if volume.VolumeSource.PersistentVolumeClaim != nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getDynamicallyProvisionedUnretainedPvs(ctx context.Context, c client.Client) (*corev1.PersistentVolumeClaimList, *corev1.PersistentVolumeList, error) {
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	if err := c.List(ctx, pvcList); err != nil {
+		return nil, nil, fail.KubeClient(err, "failed to list PVCs from user cluster.")
+	}
+	allPVList := &corev1.PersistentVolumeList{}
+	if err := c.List(ctx, allPVList); err != nil {
+		return nil, nil, fail.KubeClient(err, "failed to list PVs from user cluster.")
+	}
+	pvList := &corev1.PersistentVolumeList{}
+	for _, pv := range allPVList.Items {
+		// Check only dynamically provisioned PVs with delete reclaim policy to verify provisioner has done the cleanup
+		// this filters out everything else because we leave those be
+		if pv.Annotations[AnnDynamicallyProvisioned] != "" && pv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimDelete {
+			pvList.Items = append(pvList.Items, pv)
+		}
+	}
+
+	return pvcList, pvList, nil
+}
+
+func WaitCleanUpVolumes(ctx context.Context, logger logrus.FieldLogger, c client.Client) error {
+	logger.Infoln("Waiting for all dynamically provisioned and unretained volumes to get deleted...")
+
+	return wait.PollUntilContextTimeout(ctx, 5*time.Second, 5*time.Minute, false, func(ctx context.Context) (bool, error) {
+		pvcList, pvList, err := getDynamicallyProvisionedUnretainedPvs(ctx, c)
+		if err != nil {
+			return false, nil
+		}
+
+		if (pvcList != nil && pvList != nil) && len(pvcList.Items) == 0 && len(pvList.Items) == 0 {
+			return true, nil
+		}
```

Comment:
since we don't do anything useful with `pvcList` and `pvList` besides checking their length, I'm pretty sure we can return this info directly